### PR TITLE
Cara & Emilce -- Scrabble -- Octos

### PR DESCRIPTION
**Pull request by emilcecarlisa**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |Ability to merge and reconcile changes. Working remotely is made easy.|
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |So that it can be updated by multiple instances|
Describe an example of a test you wrote for a _nominal case_. |Tested the score of word.|
Describe an example of a test you wrote for an _edge case_. |Tested if we received nil.|
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used max and each with object method. It was helpful being concise and implicitly returning objects.|
What was one method you and your pair used to debug code? |We used puts statements and pry ocassionally. |
Is there a specific piece of code you'd like feedback on? |The draw tiles method in the tiles file.|
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |We discussed communication styles and working on specific tasks.|

### CODE REVIEW

File: lib/tilebag.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    attr_reader :tile_bag
+
+    BAG = { A: 9,	N: 6, B: 2,	O: 8, C: 2, P: 2, D: 4,	Q: 1, E: 12, R: 6, F: 2,	S: 4, G: 3, T: 6, H: 2,	U: 4, I: 9,	V: 2, J: 1,	W: 2, K: 1,	X: 1, L: 4,	Y:2, M: 2,	Z: 1 }
+
+    def initialize
+      #creates array which takes (letter, value) as index & tilebag as iterator
+      @tile_bag = BAG.each_with_object([]) { |(letter, value),tile_bag|
+      #iterate through the letters val times and shovel the
+      #letter to the individual letter tilebag
+        value.times { (tile_bag << letter.to_s)}} #
```
Comment (CheezItMan) [inline lib/tilebag.rb:13]: small indentation error.  The above loops also look awkward.  They are however an awesome way to turn the hash above into an array.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,115 @@
 module Scrabble
+
   class Scoring
+    SCORING_RUBRIK =  {
```
Comment (CheezItMan) [inline lib/scoring.rb:4]: Good use of a constant, and a good choice of data structure!

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,115 @@
...
+        highest_word = array_of_words.first
+        array_of_words.each do |word|
+          score = self.score(word)
+          if score > highest_score
+            highest_word = word
+            highest_score = score
+
+          elsif score == highest_score
+            # highest_word =
+            self.break_tie(highest_word, word)
```
Comment (CheezItMan) [inline lib/scoring.rb:71]: Nice, a helper method!

File: lib/scoring.rb [hunk 4]
```diff
@@ -1,9 +1,115 @@
...
+        puts "#{highest_word}"
+        return highest_word
+
+      end
+      #return highest_scoring_word
+
+    end
+
+    def self.break_tie(incumbent, challenger)
+
```
Comment (CheezItMan) [inline lib/scoring.rb:83]: Nicely done here

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -41,22 +44,49 @@
...
 
     it 'returns the only word in a length-1 array' do
+      words = []
+      words << "pearl"
+      Scrabble::Scoring.highest_score_from(words).must_equal "pearl"
     end
 
     it 'returns the highest word if there are two words' do
+      words = []
+      words.push("zebra", "otter")
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:64]: I suggest also varying the order of the words since that makes a difference in a tie.

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+
+    def plays #(word)
+      return @words_played
+      # @words_played.push(word)
+      # return @words_played
+    end
+
+    def play(word)
```
Comment (CheezItMan) [inline lib/player.rb:24]: Problem here, even if I've won it pushes the new word into `@words_played`

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        return false
+      else
+        return Scrabble::Scoring.score(word)
+        # self.plays(word)
+        # return a
+      end
+    end
+
+    def total_score()
+      word_scores = @words_played.map do |word|
```
Comment (CheezItMan) [inline lib/player.rb:37]: Good practice with `.map` and `.inject`

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      if total_score > 100
+        has_won = true
+      else
+        has_won = false
+      end
+      return has_won
+    end
+
+    def highest_word_score
+      highest_word = Scrabble::Scoring.highest_score_from(@words_played)
```
Comment (CheezItMan) [inline lib/player.rb:60]: Couldn't you just use the `highest_scoring_word` here?

File: specs/player_spec.rb [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    player_d = Scrabble::Player.new('player_d')
+
+    player_d.play('zzzzzz')
+    player_d.play('zzzzzz')
+
+    player_d.play('pie').must_equal false
+
+  end
+
+  it "returns score of a play/word " do
```
Comment (CheezItMan) [inline specs/player_spec.rb:42]: What about a test for what happens if you keep playing after you won?

File: lib/tilebag.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      end
+
+      puts "#{@tile_bag}"
```
Comment (CheezItMan) [inline lib/tilebag.rb:16]: puts is unneeded here:

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check
Both Teammates contributed to the codebase |  Both members committed!
Regular Commits with meaningful commit messages | Good number of commit, You should focus less on the waves and numbered tests passed and more on the functionality you added in your commit messages.
**Scoring class** |
`score` method | Check
Uses appropriate data structure to store the letter score | Check
Appropriately handles edge cases | Check
`highest_score_from` method |  Check
Appropriately handles edge cases | Check
Test implementations for `highest_score_from`| Check
**Player class** |
Uses an array to keep track of words played | Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  Check, but I left one note. 
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` | Yes, except for the case where I user keeps playing after they've won (what happens to the list of words played?)
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Neat use of Enumerables to turn the hash into an array.
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count | Check
Tests `draw_tiles` edge cases | **MISSING** Only one test for `TileBag`
Summary | Overall nice work, it looks like you ran out of time for TileBag testing, but overall very well done.
